src/index.ts

```diff
@@ -1,26 +1,71 @@
+interface DetailedEncryptionResult {
+  vault: string;
+  extractedKeyString: string;
+}
+
 interface EncryptionResult {
   data: string;
   iv: string;
   salt?: string;
 }
 
+interface DetailedDecryptResult {
+  extractedKeyString: string;
+  vault: unknown;
+  salt: string;
+}
+
+const EXPORT_FORMAT = 'jwk';
+const DERIVED_KEY_FORMAT = 'AES-GCM';
+const STRING_ENCODING = 'utf-8';
+
 /**
  * Encrypts a data object that can be any serializable value using
  * a provided password.
  *
  * @param {string} password - password to use for encryption
  * @param {R} dataObj - data to encrypt
+ * @param {CryptoKey} key - a CryptoKey instance
+ * @param {string} salt - salt used to encrypt
  * @returns {Promise<string>} cypher text
  */
-async function encrypt<R>(password: string, dataObj: R): Promise<string> {
-  const salt = generateSalt();
+async function encrypt<R>(
+  password: string,
+  dataObj: R,
+  key?: CryptoKey,
+  salt: string = generateSalt(),
+): Promise<string> {
+  const cryptoKey = key || (await keyFromPassword(password, salt));
 
-  const passwordDerivedKey = await keyFromPassword(password, salt);
-  const payload = await encryptWithKey(passwordDerivedKey, dataObj);
+  const payload = await encryptWithKey(cryptoKey, dataObj);
   payload.salt = salt;
+
   return JSON.stringify(payload);
 }
 
+/**
+ * Encrypts a data object that can be any serializable value using
+ * a provided password.
+ *
+ * @param {string} password - password to use for encryption
+ * @param {R} dataObj - data to encrypt
+ * @returns {Promise<DetailedEncryptionResult>} object with vault and extractedKeyString
+ */
+async function encryptWithDetail<R>(
```

Comment:
Rolling this back into `encrypt` might make mobile compatibility easier to manage.  It'd be a breaking change, but we can easily update mobile to return `{ vault }` rather than `vault`, and that's the only change we'd need. Whereas the alternative might be to severely complicate KeyringController to handle the two different types of encryptor libraries.

Reply:
On second thought, I am less sure of this suggestion now. Maybe we should figure out what makes sense on KeyringController first, then come back to this.